### PR TITLE
Podman machine resets all providers

### PR DESCRIPTION
**Pull request by ashley-cui**
Podman machine reset now removes and resets machines from all providers availabe on the platform.

On windows, if the user is does not have admin privs, machine will only reset WSL, but will emit a warning that it is unable to remove hyperV machines without elevated privs.

#### Does this PR introduce a user-facing change?

```release-note
The `podman machine reset` command now removes machines from all providers. 
```

### CODE REVIEW

**Comment by ashley-cui** [comment]
Unclear how to test, I don't think we have a way of creating machines with different providers in CI? Going to put no tests needed for now, if anyone disagrees and has an idea for tests, let me know.

**Comment by ashley-cui** [comment]
Ready for review, @containers/podman-maintainers  PTAL

File: cmd/podman/machine/reset.go [hunk 1]
```diff
@@ -3,18 +3,12 @@
...
-	"fmt"
-	"os"
-	"strings"
-
 	"github.com/containers/common/pkg/completion"
 	"github.com/containers/podman/v5/cmd/podman/registry"
 	"github.com/containers/podman/v5/cmd/podman/validate"
 	"github.com/containers/podman/v5/pkg/machine"
-	"github.com/containers/podman/v5/pkg/machine/env"
+	provider2 "github.com/containers/podman/v5/pkg/machine/provider"
```
Comment (Luap99) [inline cmd/podman/machine/reset.go:10]: Why is this renamed?

File: pkg/machine/shim/host.go [hunk 2]
```diff
@@ -636,38 +637,110 @@ func confirmationMessage(files []string) {
...
 		if err != nil {
-			resetErrors = multierror.Append(resetErrors, err)
+			return err
 		}
-		_, genericRm, err := mc.Remove(false, false)
+		toRemove = append(toRemove, pm)
+	}
+
+	if !opts.Force {
+		resetConfirmationMessage(toRemove)
```
Comment (Luap99) [inline pkg/machine/shim/host.go:652]: Why did we move all the printing code here? Style wise I think we should avoid as much as possible printing in the "backend" code. I know machine already ignores this elsewhere but I like it much more to have the prompts/prints under cmd/podman?

File: pkg/machine/provider/platform_windows.go [hunk 3]
```diff
@@ -43,6 +43,18 @@ func Get() (vmconfigs.VMProvider, error) {
 	}
 }
 
+func GetAll() ([]vmconfigs.VMProvider, error) {
+	providers := []vmconfigs.VMProvider{
+		new(wsl.WSLStubber),
+	}
+	if !wsl.HasAdminRights() {
+		logrus.Warn("managing hyperv machines require admin authority.")
+	} else {
```
[the remark this answers is not in the document]
Comment (mheon) [reply pkg/machine/provider/platform_windows.go:52]: I like the message, though. Drop to a Debug?

File: pkg/machine/shim/host.go (hunk 2 above)
Comment (ashley-cui) [reply pkg/machine/shim/host.go:652]: There was a precedent in Remove() so I did this mostly to avoid doing the work of looking for the machines twice - if this is done in cmd/podman, we'd have to loop through the providers and machines there to get the names, and again in Reset. Though this shouldn't be too expensive, so if you're okay with  the code dup, I can do that? The other option to avoid this would be to have Reset take the providerMachines struct as an arg, which I was initially reluctant to do since it's not a standard arg like the other functions in shim/host.go.

Let me know if either of these sounds better, or to leave it as is.

File: pkg/machine/provider/platform_windows.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashley-cui) [reply pkg/machine/provider/platform_windows.go:52]: I'm concerned that people would expect reset to hard reset, including removing HyperV machines. This is actually tricky, because doing a reset the way this is implemented right now means that when reset is run without admin rights, HyperV machine configs and files will be deleted but will not be unregistered from the hypervisor, since the machine conf dirs are deleted anyway. I also considered making `podman machine reset` an admin-only command to get around this problem but that would exclude WSL users who don't have admin access.

I can drop it to a debug, but I guess there's other issues here to think about, since nothing is perfect right now due to HyperV's perm limitations.

File: cmd/podman/machine/reset.go (hunk 1 above)
Comment (ashley-cui) [reply cmd/podman/machine/reset.go:10]: Copy and paste from other file, I can change it.

Comment (ashley-cui) [reply cmd/podman/machine/reset.go:16]: Turns out provider is already declared in machine.go, so this has to remain provider2

File: pkg/machine/provider/platform_windows.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashley-cui) [reply pkg/machine/provider/platform_windows.go]: Updated to only warn if not -f, since I think this is important information due to dangling hyperV machines.

**Comment by ashley-cui** [comment]
Hmm, needs https://github.com/containers/podman/pull/23118 to merge, then I'll rebase.

File: pkg/machine/shim/host.go (hunk 2 above)
Comment (ashley-cui) [reply pkg/machine/shim/host.go]: Moved to cmd/podman!

**Comment by ashley-cui** [comment]
@Luap99 PTAL again :)

**Review by Luap99 — COMMENTED** [review]
code wise LGTM, although I like to test this on my mac with applehv/libkrun as there is not CI test for this (I don't see a easy way to add one so that is fine). Will try to test this tomorrow morning.

**Comment by ashley-cui** [comment]
Tests passed, ready to go. @containers/podman-maintainers PTAL
